cfme/cloud/provider/__init__.py

```diff
@@ -309,41 +214,45 @@ def step(self, *args, **kwargs):
         sel.click(InfoBlock.element('Relationships', 'Images'))
 
 
-def get_all_providers(do_not_navigate=False):
+def get_all_providers():
     """Returns list of all providers"""
-    if not do_not_navigate:
-        navigate_to(CloudProvider, 'All')
-    providers = set([])
-    link_marker = "ems_cloud"
-    for page in paginator.pages():
-        for title in sel.elements("//div[@id='quadicon']/../../../tr/td/a[contains(@href,"
-                "'{}/show')]".format(link_marker)):
-            providers.add(sel.get_attribute(title, "title"))
-    return providers
-
-
-def discover(credential, cancel=False, d_type="Amazon"):
+    view = navigate_to(CloudProvider, 'All')
+    return [item.name for item in view.items.get_all(surf_pages=True)]
+
+
+def discover(credential, cancel=False, d_type="Amazon EC2"):
     """
     Discover cloud providers. Note: only starts discovery, doesn't
     wait for it to finish.
 
     Args:
       credential (cfme.base.credential.Credential):  Amazon discovery credentials.
       cancel (boolean):  Whether to cancel out of the discover UI.
+      d_type: provider name which will be discovered
     """
-    navigate_to(CloudProvider, 'Discover')
-    form_data = {'discover_select': d_type}
-    if credential:
-        form_data.update({'username': credential.principal,
-                          'password': credential.secret,
-                          'password_verify': credential.verify_secret})
-    fill(discover_form, form_data,
-         action=form_buttons.cancel if cancel else discover_form.start_button,
-         action_always=True)
+    view = navigate_to(CloudProvider, 'Discover')
+    view.fill({'discover_type': d_type})
+    if d_type == 'Amazon EC2':
```

Comment:
I want to avoid provider specific behaviour in a common method, we should more move these out as helper functions on the providers, this way we don't need to modify the core to support a new type of provider.